common/src/main/java/tachyon/security/PlainSaslServer.java

```diff
@@ -0,0 +1,139 @@
+ * the License.
+ */
+
+package tachyon.security;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import javax.security.auth.callback.Callback;
+import javax.security.auth.callback.CallbackHandler;
+import javax.security.auth.callback.NameCallback;
+import javax.security.auth.callback.PasswordCallback;
+import javax.security.auth.callback.UnsupportedCallbackException;
+import javax.security.sasl.AuthorizeCallback;
+import javax.security.sasl.SaslException;
+import javax.security.sasl.SaslServer;
+
+/**
+ * Because the Java SunSASL provider doesn’t support the server-side PLAIN mechanism.
+ * There is a new provider needed to register to support server-side PLAIN mechanism.
+ * There are three basic steps in implementing a SASL security provider:
+ * 1.Write a class that implements the SaslServer interface
+ * 2.Write a factory class implements the SaslServerFactory
+ * 3.Write a JCA provider that registers the factory
+ */
+public class PlainSaslServer implements SaslServer {
+  private String mUser;
+  private final CallbackHandler mHandler;
+
+  PlainSaslServer(CallbackHandler mHandler, String authMethodStr) throws SaslException {
+    this.mHandler = mHandler;
+  }
+
+  @Override
+  public String getMechanismName() {
+    return "PLAIN";
+  }
+
+  @Override
+  public byte[] evaluateResponse(byte[] response) throws SaslException {
+    try {
+      // parse the response
+      // message   = [authzid] UTF8NUL authcid UTF8NUL passwd'
+
+      Deque<String> tokenList = new ArrayDeque<String>();
+      StringBuilder messageToken = new StringBuilder();
+      for (byte b : response) {
+        if (b == 0) {
+          tokenList.addLast(messageToken.toString());
+          messageToken = new StringBuilder();
+        } else {
+          messageToken.append((char) b);
+        }
+      }
+      tokenList.addLast(messageToken.toString());
+
+      // validate response
+      if (tokenList.size() < 2 || tokenList.size() > 3) {
+        throw new SaslException("Invalid message format");
+      }
+      String passwd = tokenList.removeLast();
+      mUser = tokenList.removeLast();
+      // optional authzid
+      String authzId;
+      if (tokenList.isEmpty()) {
+        authzId = mUser;
+      } else {
+        authzId = tokenList.removeLast();
+      }
+      if (mUser == null || mUser.isEmpty()) {
+        throw new SaslException("No user name provided");
+      }
+      if (passwd == null || passwd.isEmpty()) {
+        throw new SaslException("No password name provided");
+      }
+
+      NameCallback nameCallback = new NameCallback("User");
+      nameCallback.setName(mUser);
+      PasswordCallback pcCallback = new PasswordCallback("Password", false);
+      pcCallback.setPassword(passwd.toCharArray());
+      AuthorizeCallback acCallback = new AuthorizeCallback(mUser, authzId);
+
+      Callback[] cbList = {nameCallback, pcCallback, acCallback};
+      mHandler.handle(cbList);
```

Comment:
Thanks for your question. These three callbacks will be used to do some verify. You can see my design doc about the simple authentication (https://docs.google.com/document/d/1Yjjm0DtdWKTd0hJnYbYOJQTKT9Fpe7gVtp1nGZNRadQ). section[4.3.2.1]. These callbacks will be used with the authenticatonProvider. Thanks